enterprise_access/apps/content_assignments/api.py

```diff
@@ -383,6 +373,118 @@ def _deduplicate_learner_emails_to_allocate(learner_emails):
     return deduplicated
 
 
+def _map_allocation_emails_with_lms_user_ids(learner_emails_to_allocate):
+    """
+    To allocate assignments, we'll need to lookup existing assignments
+    by both email *and* lms_user_id. We'll also use these lms_user_ids
+    to populate the `lms_user_id` field on existing assignments that
+    don't currently have the field populated.
+
+    We'll also need a reverse lookup of lms_user_id -> email
+    to re-populate the `learner_email` of existing assignments that are
+    in an expired state (and thus have an auto-generated, non-identifiable `learner_email` value).
+
+    Returns:
+      Two dicts, the first mapping email -> lms_user_id, and the second
+      mapping lms_user_id -> email.
+    """
+    lms_user_ids_by_email = _get_lms_user_ids_by_email(learner_emails_to_allocate)
+    emails_by_lms_user_id = {
+        lms_user_id: email for email, lms_user_id in lms_user_ids_by_email.items()
+    }
+    return lms_user_ids_by_email, emails_by_lms_user_id
+
+
+def _get_lms_user_ids_by_email(emails):
+    """
+    Helper to return a mapping of learner email addresses to lms_user_id.
+    If no user record exists with a given email address, it will *not*
+    be present in the mapping.
+
+    Performance note: This results in one or more reads against the User model.
+      The chunk size has been tuned to minimize the number of reads
+      while simultaneously avoiding hard limits on statement length.
+    """
+    lms_user_ids_by_email = {}
+    for email_chunk in chunks(emails, USER_EMAIL_READ_BATCH_SIZE):
+        # Construct a list of tuples containing (email, lms_user_id) for every email in this chunk.
+        # this is the part that could exceed max statement length if batch size is too large.
+        # There's no case-insensitive IN query in Django, so we have to build up a big
+        # OR type of query.
+        email_lms_user_id = User.objects.filter(
+            _inexact_email_filter(email_chunk, field_name='email'),
+            lms_user_id__isnull=False,
+        ).values_list('email', 'lms_user_id')
+
+        # dict() on a list of 2-tuples treats the first elements as keys and second elements as values.
+        lms_user_ids_by_email.update(dict(email_lms_user_id))
+    return lms_user_ids_by_email
+
+
+def _get_existing_assignments_for_allocation(
+        assignment_configuration, learner_emails_to_allocate, content_key, lms_user_ids_by_email,
+):
```

Comment:
Here's another big change - we now search for existing assignments by either the client-provided emails, or by any existing lms_user_ids that are associated with any of the provided emails.